### ISSUE TITLE
Hiding private platforms from the main repository

### ISSUE THREAD

**Issue by mrhelmut** [MonoGame/MonoGame#6947]
Hi there!

Due to confidentiality, it would be cool if private platform names disappear from the public repository (that is, PlayStation4, XBoxOne, Switch, and PSVita), while still maintaining an architecture that would allow the content pipeline to be populated with those platforms and build without having to modify the public's repo' code. As well as having the content builder to check if the platform is properly set and supported.

We can always use code names for unreleased/unannounced platforms, but that's quite touchy from a legal standpoint.

Private platforms are repositories intended to the cloned as subdirectories to the main public repository. For instance, here's how the folders are discovered currently:
https://github.com/MonoGame/MonoGame/blob/5f6a01a34e6f677b7f42b15d544e75ebf6d07fee/MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj#L227-L252

E.g. here are the references to PlayStation 4:
https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework.Content.Pipeline/TargetPlatform.cs#L87
https://github.com/MonoGame/MonoGame/blob/e4d4e9cbe140cf26d19467cf37e3dcece0c28ec9/MonoGame.Framework/Utilities/MonoGamePlatform.cs#L60
https://github.com/MonoGame/MonoGame/blob/e4d4e9cbe140cf26d19467cf37e3dcece0c28ec9/MonoGame.Framework/Utilities/PlatformInfo.cs#L38
https://github.com/MonoGame/MonoGame/blob/cf275525fc022a92a9e5905ed500c1f810820707/MonoGame.Framework.Content.Pipeline/MonoGame.Content.Builder.targets#L56
https://github.com/MonoGame/MonoGame/blob/5f6a01a34e6f677b7f42b15d544e75ebf6d07fee/MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj#L240
https://github.com/MonoGame/MonoGame/blob/1404ab7334f3d303a60b576eaa5d1153e537d27a/Tools/MonoGame.Content.Builder/MonoGame.Content.Builder.targets#L231
https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework.Content.Pipeline/Serialization/Compiler/ContentWriter.cs#L55
https://github.com/MonoGame/MonoGame/blob/431271a92c3ebc2a6b6811fde09376865b512240/MonoGame.Framework/Content/ContentManager.cs#L49

```MonoGame.Framework.Content.Pipeline.csproj``` can be fixed with some wildcards, but platform identifiers within the content builder code will require some refactoring and new mechanisms to build including private repo', identify the platforms and validate that the targeted platform is supported by the current build of the content pipeline.

Ideas and insights welcome to refactor this! :)

**Comment by Jjagg** [comment]
For enums we can turn them into partial classes with members for what are now the enum fields e.g. `public static readonly string DesktopGL = "DesktopGL";`.

For the runtime we can use partial classes.
For the Content Pipeline we can maybe do a reflection-based thing with attributes? So something like `[ExportContentPlatform("PlayStation4", 'P')]`.

For the .targets file, I'm not sure what we should do. An easy thing to do, but not a great fix is to use a fixed prefix for new platforms and then just check if `MonoGamePlatform` starts with that prefix. Something like `PPLAT_*` so we'd have for example `PPLAT_PlayStation5`.
There's not really a way to modify or extend it, so if we want a 'correct' fix we'd need to cut out the proprietary platforms from the .targets file in the public repo and have different .targets file for each proprietary platform. We can generate those with a t4 template or something. We might not be able to make that as user-friendly though.

**Comment by mrhelmut** [comment]
For the targets, we could simply check if the platform is set, and leave it be. Then having mgcb to throw an error if the platform is unsupported.

**Comment by Jjagg** [comment]
Yeah, that works. Users still get an error at build time, so it's not that big a deal.

**Comment by cra0zy** [comment]
Instead I would rather use 1 single "PrivatePlatform" enum that gets shared between all the private repos. That way if you build MG against switch repo, that enum now means Switch, however if you build it against Xbox repo that enum now means Xbox.

**Comment by tomspilman** [comment]
@cra0zy - Often we have one copy of the MG repo and use that to target multiple console platforms.  This includes having one copy of the Pipeline tool and building content for multiple console platforms.  So ideally we could have support for multiple private platforms at once.

**Comment by mrhelmut** [comment]
Another issue would be that the content of all consoles will be output to the same directory, which is something to avoid when you build multiple platforms at once.

**Comment by fdrobidoux** [comment]
I'm mostly wondering what the root issue is : What makes those platforms require "confidentiality"? And what do you mean by "private platforms"? Is it because access to their SDK is on an on-demand basis instead of being straight up open to the public?

**Comment by Jjagg** [comment]
For cases where we have an implementation that people can't now about (not publicly announced platform) we want to be able to build without too much hassle, but without requiring any reference to the platform in the public repo.

**Comment by fdrobidoux** [comment]
That makes sense. Carry on.
